Use elevation to implement shadows on Android

corbt opened the pull request:
This PR includes a working interface to Android's `elevation` property implemented as an RN style. Elevation is the only (easy) [platform-supported way to create shadows](http://developer.android.com/training/material/shadows-clipping.html). For it to work note that you must be running on Android 5.0+, and add `elevation` to a view with a `backgroundColor` set. These are platform limitations.

This PR is not intended to be merged in its current state, but rather to inform the discussion from #2768. At a minimum, before merging we would need to add the elevation style to the docs and rebase this to master (**EDIT** I have now rebased on master because from v0.14.2 too many commits were being pulled in -- haven't tested it since the rebase though). Additionally, it might be good to add tests, although I couldn't find any for the Android code. I'm happy to get that done if this feature gets signed off by the React Native team.

Finally, as I argued in #2768 I think that `elevation` is a useful abstraction over manual shadow management, and I would like to see it ported over to the iOS side as well. This would allow cross-platform developers to create shadows with a single `elevation` style. The alternative would be to just keep the shadow* APIs on iOS and elevation for Android, which to me feels less clean. However, this could be merged before that work is done, and the cross-platform implementation can come later.

Example usage:

```jsx
var RNelevation = React.createClass({
  render: function() {
    return (
      <View style={styles.container}>
        <Text style={styles.title}>
          React Native Elevation Test
        </Text>
        <View style={styles.bigBox}>
          <View style={styles.smallBox} />
        </View>
      </View>
    );
  }
});

var styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
  title: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  bigBox: {
    width: 200,
    height: 200,
    elevation: 20,
    backgroundColor: 'green'
  },
  smallBox: {
    backgroundColor: 'red',
    width: 100,
    height: 100,
    elevation: 10
  }
});
```

Result:
![rn elevation](https://cloud.githubusercontent.com/assets/176426/11214434/2878f598-8d39-11e5-8d3c-a117c22ea348.png)

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java:
```diff
@@ -123,6 +126,20 @@ public int getOpacity() {
+              : 0;
+
+      outline.setRoundRect(getBounds(), borderRadius+extraRadiusFromBorderWidth);
+    }
```
Can you please add a comment explaining why this is needed? Why is it needed? Isn't setting `elevation` enough?

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java:
```diff
@@ -123,6 +126,20 @@ public int getOpacity() {
+              ? mBorderWidth.get(Spacing.ALL)/2
+              : 0;
+
+      outline.setRoundRect(getBounds(), borderRadius+extraRadiusFromBorderWidth);
```
nit: spaces around +

corbt replied:
sorry, not sure what this comment means?

corbt replied:
ok, figured it out.

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewManager.java:
```diff
@@ -64,6 +65,11 @@ public void setBorderStyle(ReactViewGroup view, @Nullable String borderStyle) {
 
+  @ReactProp(name = "elevation")
+  public void setElevation(ReactViewGroup view, float elevation) {
+    ViewCompat.setElevation(view, PixelUtil.toPixelFromDIP(elevation));
```
Why not check for platform level and call `View.setElevation`? http://developer.android.com/reference/android/view/View.html#setElevation(float)?

corbt replied:
I suspect that's essentially what `ViewCompat.setElevation` does, with the addition that if Android ever does backport `elevation` to earlier APIs (admittedly unlikely) it'll start working there as well. But yeah, doing it the other way would work fine too. Is there a reason to prefer one style over the other?

mkonicek replied:
You're right. OK, let's keep this one.

http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/5.1.1_r1/android/support/v4/view/ViewCompat.java#ViewCompat.setElevation%28android.view.View%2Cfloat%29

e.g. the JellyBean implementation is empty:
http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/5.1.1_r1/android/support/v4/view/ViewCompat.java#698

mkonicek commented:
> Additionally, it might be good to add tests, although I couldn't find any for the Android code.

Absolutely, we plan to open source the Android integration tests.

mkonicek commented:
Totally agree we can add `elevation` now and decide on a cross-platform API for shadows later. One thing that's clear is that we shouldn't be trying to infer the elevation from the shadow properties. Therefore starting by adding `elevation` is a good way to go.

mkonicek commented:
Thanks for much for working on this! :+1:

I've read briefly through https://github.com/facebook/react-native/issues/2768 but the discussion is quite long, I think it'll be helpful to discuss directly here now we know adding an `elevation` prop is the way to go :)

corbt commented:
Ok, I've pushed some changes related to @mkonicek's review. As far as why `getOutline` needs to be implemented: unfortunately I haven't been able to find any documentation that explains the requirement (or even the `setElevation` source code that might point to where the outline is used). But intuitively, it makes sense that to cast a shadow around an object you'd need to know where its bounds are. And practically speaking, it's there because when I get rid of it the shadow stops working. :) Anyway, I added a brief comment mentioning its relationship to `elevation`.

corbt commented:
I should mention that this PR only adds `elevation` to the `View` object, nothing else. Would it make sense to add explicit support for it to other UI components, or just require people to wrap anything they want shadows on inside a `View`? How does it work right now on iOS for shadows?

mkonicek commented:
Thanks for adding the comment to `getOutline`! Just to be safe, have you tested what setting `elevation` does on API < 21? It should have no effect.

> it appears that the border* styles on Android are only supported on Views and ImageViews.

Yes this is correct. Yes, adding elevation to those two (and even better starting just with `View` in this PR) sounds good.

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java:
```diff
@@ -123,6 +126,21 @@ public int getOpacity() {
+      float borderRadius = (!CSSConstants.isUndefined(mBorderRadius) && mBorderRadius > 0)
+              ? mBorderRadius
+              : 0;
+      float extraRadiusFromBorderWidth = (mBorderWidth != null)
```
Is this needed? How does the shadow look without it? Maybe we just need higher elevation?

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java:
```diff
@@ -123,6 +126,21 @@ public int getOpacity() {
+              ? mBorderRadius
+              : 0;
+      float extraRadiusFromBorderWidth = (mBorderWidth != null)
+              ? mBorderWidth.get(Spacing.ALL)/2
```
What if there's only the bottom border for example?

mkonicek commented:
Can you please provide some screenshots for views with thick borders / round corners / thick border only on one side etc.?

brentvatne commented:
This is fantastic! Looking forward to getting it in, thanks @corbt and @mkonicek! :smile: :rocket:

corbt commented:
@mkonicek ok I'm getting distracted by some real-world work :wink:; I'll get those screenshots for you but probably on Thursday.

mkonicek commented:
Haha no problem @corbt! I was curious about the `extraRadiusFromBorderWidth` and `mBorderWidth.get(Spacing.ALL)`.

corbt commented:
Ok, find some screen shots of different border/radius combinations below.

I've also verified that nothing happens on API < 21 (checked in an emulator on API 16), and updated the PR to fix a bug where there is a large border and no border radius.

![device-2015-11-19-160801](https://cloud.githubusercontent.com/assets/176426/11276074/77380714-8ed8-11e5-98ff-07c7a8a9b700.png)

Code for the above:

```jsx
class Example extends Component {
  render() {
    return <View style={styles.exampleContainer}>
      <View>
        <Text style={styles.subtitle}>{JSON.stringify(this.props.style, null, 2).slice(1, -1)}</Text>
      </View>
      <View style={[styles.example, this.props.style]} />
    </View>;
  }
}
        // <View style={{flexDirection: 'row'}}>

var RNelevation = React.createClass({
  render: function() {
    return (
      <View style={styles.container}>
        <Example style={{borderBottomWidth: 30}} />
        <Example style={{borderRadius: 45}} />
        <Example style={{borderRadius: 30, borderWidth: 30}} />
        <Example style={{borderWidth: 5, borderBottomWidth: 30}} />
        <Example style={{borderBottomWidth: 30, borderRadius: 30}} />
        <Example style={{borderWidth: 30, borderRadius: 1}} />
        <Example style={{borderWidth: 30}} />
      </View>
    );
  }
});

var styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'center',
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  exampleContainer: {
    alignItems: 'center',
    width: 200,
    margin: 20,
  },
  example: {
    width: 100,
    height: 100,
    backgroundColor: 'red',
    elevation: 10,
    margin: 50,
    marginTop: 5,
    borderColor: 'green'
  },
  subtitle: {
    textAlign: 'center',
  }
});
```

mkonicek commented:
Thanks for the explanation and the screenshots! We should add this an an example to the UI Explorer.

@facebook-github-bot shipit

satya164 commented:
Another property of `elevation` seems to be that it changes the stack order. An element with `elevation` value 6 will be always above an element with `elevation` value 4.

That's one caveat if we want to have the same thing on iOS, as RN doesn't support changing the stack order.

ide commented:
We should mention @satya164's point in a docblock for this property.

corbt commented:
I'm happy to add documentation, including @satya164's important observation that `elevation` affects stacking order. In fact I was planning on adding documentation to this PR before it was merged, as soon as there was a consensus that this was a good idea. :) Not sure about the mechanics of @facebook-github-bot now that this has been "shipped", should I still add that to this PR or open a new one?

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java:
```diff
@@ -123,6 +126,23 @@ public int getOpacity() {
+      outline.setRoundRect(getBounds(), mBorderRadius + extraRadiusFromBorderWidth);
+    }
+    else {
+      outline.setRect(getBounds());
```
nit: call `super.getOutline`: https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/graphics/java/android/graphics/drawable/Drawable.java#989

mkonicek commented:
It hasn't landed yet due to Buck not seeing `@NonNull. We use Buck to build the code internally. I could add that to the Buck file but suggest we just remove the annotation and import here. It makes the dependencies simpler and we don't use that annotation (even though here you're overriding the method so you technically should but well..), we just use `javax.annotation.Nullable` - we assume everything else is non-null by default.

Once it lands this PR would be automatically closed. You can still edit this PR and I'll comment 'shipit' again.

For details on how the bot works you can check slides 45+ in this presentation: https://speakerdeck.com/mkonicek/under-the-hood-of-react-native

Adding docs about the z-order is a great idea!

corbt commented:
@mkonicek ok, removed the `@NotNull`, added the relevant call to `super`, and added some documentation. Unfortunately it looks like there isn't any way to expose docblocks for styles currently (which I guess makes sense for styles pulled directly from CSS, but would be nice for less obvious ones like this), but I tweaked "Known Issues" to describe this solution, and added a docblock to the styles page anyway in case rendering code for those is ever added to doc generation.

mkonicek commented:
Cool, nice work!

@facebook-github-bot shipit

mkonicek commented:
Landing this now. Due to a limitation of how the code import script works we'll have to commit the changes to the docs separately once this lands. I'm happy to do that for you.

satya164 commented:
@mkonicek Will it go in 0.16?

mkonicek commented:
Let's cherry-pick it.

mkonicek commented:
@corbt This is finally in :tada: I've made a few small tweaks while landing it.

mkonicek commented:
Updated Known Issues. Don't think no cross-platform shadows are a limitation. Shadows work differently on each platform and I believe it's okay to use the platform-specific props: https://github.com/facebook/react-native/commit/0f89696f76845925de425bbd6dd29cae8e5f0b43

You might also not want shadows to look exactly the same in an iOS and Android app.

mkonicek commented:
Thanks so much for working on this @corbt!

corbt commented:
Sweet, glad this is in! :+1:

jcsmesquita commented:
Is it possible to use this now? Which version do I need to install? Great work btw :rocket:

corbt commented:
@jcsmesquita currently it looks like it's only in master, although @mkonicek said they might cherry-pick it into 0.16. So it'll either be in the 0.16 release or 0.17. But if you want to use it right now you'll need to clone master.

jcsmesquita commented:
@corbt Got it :) thanks for the explanation!

Kishanjvaghela commented:
I have implemented CardView for react-native with elevation, that support android(All version) and iOS. Let me know is it help you or not. https://github.com/Kishanjvaghela/react-native-cardview
